File: src/readers/gtfs/schedule/locationGroupStops.ts
import { parseCSVAsRecord } from '../../';

/**
 * # Location Group Stops
 *
 * **Optional**
 * Assigns stops from `stops.txt` to location groups (`location_groups.txt`).
 */
export class GTFSLocationGroupStop {
  /**
   * **Required**
   * Identifies a location group (`location_groups.location_group_id`).
   */
  locationGroupId: string;
  /**
   * **Required**
   * Identifies a stop (`stops.stop_id`) belonging to that location group.
   */
  stopId: string;

  /** @param data - the parsed GTFS CSV data */
  constructor(data: Record<string, string>) {
    this.locationGroupId = data.location_group_id;
    this.stopId = data.stop_id;
  }
}

/**
 * @param input - the input string to parse from
 * @returns - an array of GTFSLocationGroupStops
 */
export function parseGTFSLocationGroupStops(input: string): GTFSLocationGroupStop[] {
  const data = parseCSVAsRecord(input);
  return data.map((d) => new GTFSLocationGroupStop(d));
}

File: rust/readers/gtfs/schedule/location_group_stops.rs
use crate::readers::csv::parse_csv_as_record;
use alloc::{string::String, vec::Vec};
use s2json::MValueCompatible;

/// # Location Group Stops
///
/// **Optional**
/// Assigns stops from `stops.txt` to location groups (`location_groups.txt`).
#[derive(Debug, Default, Clone, PartialEq, MValueCompatible)]
pub struct GTFSLocationGroupStop {
    /// **Required**
    /// Identifies a location group (`location_groups.location_group_id`).
    pub location_group_id: String,
    /// **Required**
    /// Identifies a stop (`stops.stop_id`) belonging to that location group.
    pub stop_id: String,
}
impl GTFSLocationGroupStop {
    /// Create a new GTFSLocationGroupStop
    pub fn new(source: &str) -> Vec<GTFSLocationGroupStop> {
        let mut res = Vec::new();
        for record in parse_csv_as_record::<GTFSLocationGroupStop>(source, None, None) {
            res.push(record);
        }
        res
    }
}
